pkg/relayer/session/session.go

```diff
@@ -134,16 +168,59 @@ func (rs *relayerSessionsManager) Start(ctx context.Context) {
 		claimedSessionsObs := rs.createClaims(ctx, supplierClient, supplierSessionsToClaimObs)
 		rs.submitProofs(ctx, supplierClient, claimedSessionsObs)
 	}
+
+	// Stop the relayer sessions manager when the context is done.
+	// This is necessary to ensure that in case of a shutdown, all the sessions trees
+	// are persisted along their root hashes.
+	go func() {
+		<-ctx.Done()
+		rs.Stop()
+	}()
+
+	return nil
 }
 
-// Stop unsubscribes all observables from the InsertRelays observable which
-// will close downstream observables as they drain.
-//
-// TODO_TECHDEBT: Either add a mechanism to wait for draining to complete
-// and/or ensure that the state at each pipeline stage is persisted to disk
-// and exit as early as possible.
+// Stop performs a complete shutdown of the relayerSessionsManager by:
+// 1. Closing connections and canceling subscriptions
+// 2. Persisting all session data to storage
+// 3. Releasing resources and clearing memory
+// This ensures no data is lost during shutdown and resources are properly cleaned up.
 func (rs *relayerSessionsManager) Stop() {
+	// Close the block client and unsubscribe from all observables to stop receiving events.
+	// While process termination would eventually clean these up, proper shutdown
+	// is important for graceful termination and testing scenarios.
+	rs.blockClient.Close()
 	rs.relayObs.UnsubscribeAll()
+
+	// Persist each active session's state to disk and properly close the associated
+	// key-value stores. This ensures that all accumulated relay data (including root
+	// hashes needed for claims) is safely stored before shutdown.
+	for _, supplierSessionTrees := range rs.sessionsTrees {
+		for _, sessionTreesAtHeight := range supplierSessionTrees {
+			for _, sessionTree := range sessionTreesAtHeight {
+				if err := rs.persistSessionMetadata(sessionTree); err != nil {
+					rs.logger.Error().Err(err).Msgf(
+						"failed to persist session metadata for sessionId %q",
+						sessionTree.GetSessionHeader().GetSessionId(),
+					)
+				}
+
+				if err := sessionTree.Stop(); err != nil {
+					rs.logger.Error().Err(err).Msgf(
+						"failed to stop session tree store for sessionId %q",
+						sessionTree.GetSessionHeader().GetSessionId(),
+					)
+				}
+			}
+		}
+	}
+
+	// Close the metadata store that tracks all sessions and release its resources.
+	// Then clear the in-memory sessions map for testing purposes.
```

Comment:
Why testing purposes?

Reply:
I only needed it on tests, In a real world scenario, the process shut-down should take care of this.

Will remove that comment anyway, as it's misleading to think that it's ONLY here for testing purposes.